File: tapo/examples/tapo_h100.rs
//! H100 Example

use log::{info, LevelFilter};
use std::env;
use std::time::Duration;
use tapo::requests::{AlarmDuration, AlarmRingtone, AlarmVolume};
use tapo::responses::ChildDeviceHubResult;
use tapo::{ApiClient, HubDevice};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let log_level = env::var("RUST_LOG")
        .unwrap_or_else(|_| "info".to_string())
        .parse()
        .unwrap_or(LevelFilter::Info);

    pretty_env_logger::formatted_timed_builder()
        .filter(Some("tapo"), log_level)
        .init();

    let tapo_username = env::var("TAPO_USERNAME")?;
    let tapo_password = env::var("TAPO_PASSWORD")?;
    let ip_address = env::var("IP_ADDRESS")?;

    let hub = ApiClient::new(tapo_username, tapo_password)
        .h100(ip_address)
        .await?;

    let device_info = hub.get_device_info().await?;
    info!("Device info: {device_info:?}");

    info!("Getting child devices...");
    let child_device_list = hub.get_child_device_list().await?;

    for child in child_device_list {
        match child {
            ChildDeviceHubResult::KE100(device) => {
                info!(
                    "Found KE100 child device with nickname: {}, id: {}, current temperature: {} {:?} and target temperature: {} {:?}.",
                    device.nickname,
                    device.device_id,
                    device.current_temperature,
                    device.temperature_unit,
                    device.target_temperature,
                    device.temperature_unit,
                );
            }
            ChildDeviceHubResult::S200B(device) => {
                let s200b = hub
                    .s200b(HubDevice::ByDeviceId(device.device_id.clone()))
                    .await?;
                let trigger_logs = s200b.get_trigger_logs(5, 0).await?;

                info!(
                    "Found S200B child device with nickname: {}, id: {}, last 5 trigger logs: {:?}.",
                    device.nickname, device.device_id, trigger_logs
                );
            }
            ChildDeviceHubResult::T100(device) => {
                let t100 = hub
                    .t100(HubDevice::ByDeviceId(device.device_id.clone()))
                    .await?;
                let trigger_logs = t100.get_trigger_logs(5, 0).await?;

                info!(
                    "Found T100 child device with nickname: {}, id: {}, detected: {}, last 5 trigger logs: {:?}.",
                    device.nickname, device.device_id, device.detected, trigger_logs
                );
            }
            ChildDeviceHubResult::T110(device) => {
                let t110 = hub
                    .t110(HubDevice::ByDeviceId(device.device_id.clone()))
                    .await?;
                let trigger_logs = t110.get_trigger_logs(5, 0).await?;

                info!(
                    "Found T110 child device with nickname: {}, id: {}, open: {}, last 5 trigger logs: {:?}.",
                    device.nickname, device.device_id, device.open, trigger_logs
                );
            }
            ChildDeviceHubResult::T300(device) => {
                let t300 = hub
                    .t300(HubDevice::ByDeviceId(device.device_id.clone()))
                    .await?;
                let trigger_logs = t300.get_trigger_logs(5, 0).await?;

                info!(
                    "Found T300 child device with nickname: {}, id: {}, in_alarm: {}, water_leak_status: {:?}, last 5 trigger logs: {:?}.",
                    device.nickname,
                    device.device_id,
                    device.in_alarm,
                    device.water_leak_status,
                    trigger_logs
                );
            }
            ChildDeviceHubResult::T310(device) | ChildDeviceHubResult::T315(device) => {
                let t31x = hub
                    .t315(HubDevice::ByDeviceId(device.device_id.clone()))
                    .await?;
                let temperature_humidity_records = t31x.get_temperature_humidity_records().await?;

                info!(
                    "Found T31X child device with nickname: {}, id: {}, temperature: {} {:?}, humidity: {}%, earliest temperature and humidity record available: {:?}.",
                    device.nickname,
                    device.device_id,
                    device.current_temperature,
                    device.temperature_unit,
                    device.current_humidity,
                    temperature_humidity_records.records.first()
                );
            }
            _ => {
                info!("Found unsupported device.")
            }
        }
    }

    info!("Triggering the alarm ringtone 'Alarm 1' at a 'Low' volume for '3 Seconds'...");
    hub.play_alarm(
        AlarmRingtone::Alarm1,
        AlarmVolume::Low,
        AlarmDuration::Seconds(3),
    )
    .await?;

    let device_info = hub.get_device_info().await?;
    info!("Is device ringing?: {:?}", device_info.in_alarm);

    info!("Stopping the alarm after 1 Second...");
    tokio::time::sleep(Duration::from_secs(1)).await;
    hub.stop_alarm().await?;

    let device_info = hub.get_device_info().await?;
    info!("Is device ringing?: {:?}", device_info.in_alarm);

    Ok(())
}

File: tapo-py/examples/tapo_h100.py
"""H100 Example"""

import asyncio
import os

from tapo import ApiClient
from tapo.requests import AlarmRingtone, AlarmVolume, AlarmDuration
from tapo.responses import KE100Result, S200BResult, T100Result, T110Result, T300Result, T31XResult


async def main():
    tapo_username = os.getenv("TAPO_USERNAME")
    tapo_password = os.getenv("TAPO_PASSWORD")
    ip_address = os.getenv("IP_ADDRESS")

    client = ApiClient(tapo_username, tapo_password)
    hub = await client.h100(ip_address)

    device_info = await hub.get_device_info()
    print(f"Device info: {device_info.to_dict()}")

    child_device_list = await hub.get_child_device_list()

    for child in child_device_list:
        if child is None:
            print("Found unsupported device.")
        elif isinstance(child, KE100Result):
            print(
                "Found KE100 child device with nickname: {}, id: {}, current temperature: {:.2f} {} and target temperature: {:.2f} {}.".format(
                    child.nickname,
                    child.device_id,
                    child.current_temperature,
                    child.temperature_unit,
                    child.target_temperature,
                    child.temperature_unit,
                )
            )
        elif isinstance(child, S200BResult):
            s200b = await hub.s200b(device_id=child.device_id)
            trigger_logs = await s200b.get_trigger_logs(5, 0)

            print(
                "Found S200B child device with nickname: {}, id: {}, last 5 trigger logs: {}.".format(
                    child.nickname,
                    child.device_id,
                    [log.to_dict() for log in trigger_logs.logs],
                )
            )
        elif isinstance(child, T100Result):
            t100 = await hub.t100(device_id=child.device_id)
            trigger_logs = await t100.get_trigger_logs(5, 0)

            print(
                "Found T100 child device with nickname: {}, id: {}, detected: {}, last 5 trigger logs: {}.".format(
                    child.nickname,
                    child.device_id,
                    child.detected,
                    [log.to_dict() for log in trigger_logs.logs],
                )
            )
        elif isinstance(child, T110Result):
            t110 = await hub.t110(device_id=child.device_id)
            trigger_logs = await t110.get_trigger_logs(5, 0)

            print(
                "Found T110 child device with nickname: {}, id: {}, open: {}, last 5 trigger logs: {}.".format(
                    child.nickname,
                    child.device_id,
                    child.open,
                    [log.to_dict() for log in trigger_logs.logs],
                )
            )
        elif isinstance(child, T300Result):
            t300 = await hub.t300(device_id=child.device_id)
            trigger_logs = await t300.get_trigger_logs(5, 0)

            print(
                "Found T300 child device with nickname: {}, id: {}, in_alarm: {}, water_leak_status: {}, last 5 trigger logs: {}.".format(
                    child.nickname,
                    child.device_id,
                    child.in_alarm,
                    child.water_leak_status,
                    [log.to_dict() for log in trigger_logs.logs],
                )
            )
        elif isinstance(child, T31XResult):
            t31x = await hub.t315(device_id=child.device_id)
            temperature_humidity_records = await t31x.get_temperature_humidity_records()

            print(
                "Found T31X child device with nickname: {}, id: {}, temperature: {:.2f} {}, humidity: {}%, earliest temperature and humidity record available: {}.".format(
                    child.nickname,
                    child.device_id,
                    child.current_temperature,
                    child.temperature_unit,
                    child.current_humidity,
                    (
                        temperature_humidity_records.records[0].to_dict()
                        if temperature_humidity_records.records
                        else None
                    ),
                )
            )

    print(f"Triggering the alarm ringtone 'Alarm 1' at a 'Low' volume for '3 Seconds'...")
    await hub.play_alarm(AlarmRingtone.Alarm1, AlarmVolume.Low, AlarmDuration.Seconds, seconds=3)

    device_info = await hub.get_device_info()
    print(f"Is device ringing?: {device_info.in_alarm}")

    print("Stopping the alarm after 1 Second...")
    await asyncio.sleep(1)
    await hub.stop_alarm()

    device_info = await hub.get_device_info()
    print(f"Is device ringing?: {device_info.in_alarm}")


if __name__ == "__main__":
    asyncio.run(main())
